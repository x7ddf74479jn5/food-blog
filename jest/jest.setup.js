// Optional: configure or set up a testing framework before each test.
// If you delete this file, remove `setupFilesAfterEnv` from `jest.config.js`

// Used for __tests__/testing-library.js
// Learn more: https://github.com/testing-library/jest-dom
import "@testing-library/jest-dom/extend-expect";

window.matchMedia = jest.fn().mockImplementation((query) => {
  return {
    addListener: jest.fn(),
    matches: false,
    media: query,
    onchange: null,
    removeListener: jest.fn(),
  };
});
