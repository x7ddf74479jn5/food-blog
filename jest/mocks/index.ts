export { mockNextImage } from "./image";
export type { NextRouter } from "./router";
export { mockRouter, RouterContext, withMockedRouter } from "./router";
export { setupWindow } from "./window";
