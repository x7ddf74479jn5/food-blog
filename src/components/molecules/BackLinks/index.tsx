import { useRouter } from "next/router";
import { FaChevronLeft } from "react-icons/fa";

import NextLink from "@/components/atoms/NextLink";

export type Props = {
  links: Array<{
    href: string;
    label: string;
  }>;
};

const BackLinks: React.FC<Props> = ({ links }) => {
  const router = useRouter();

  const handleBack = () => {
    router.back();
  };

  return (
    <ul className="flex flex-col space-y-2 align-top">
      <li>
        <button onClick={handleBack} className="flex flex-row items-center gap-2">
          <FaChevronLeft />
          前のページへ
        </button>
      </li>
      {links.map((link, index) => (
        <li key={index}>
          <NextLink href={link.href} className="flex flex-row items-center gap-2">
            <FaChevronLeft />
            {link.label}
          </NextLink>
        </li>
      ))}
    </ul>
  );
};

export default BackLinks;
