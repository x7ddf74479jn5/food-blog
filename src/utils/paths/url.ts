export const apiRoute = {
  apiPreview: "/api/preview",
  apiArticles: "/api/articles",
} as const;

export const urlTable = {
  home: "/",
  root: "/",
  articles: "/articles",
  categories: "/articles/categories",
  tags: "/articles/tags",
  preview: "/preview",
  search: "/search",
} as const;

const labelTable = {
  [urlTable.home]: "レシピ一覧へ",
  [urlTable.categories]: "カテゴリー別一覧へ",
  [urlTable.tags]: "タグ別一覧へ",
} as const;

type LabelTableKey = keyof typeof labelTable;

export const getBackLinks = (paths: LabelTableKey[]): { href: string; label: string }[] => {
  return paths.map((path) => {
    return { href: path, label: labelTable[path] };
  });
};
