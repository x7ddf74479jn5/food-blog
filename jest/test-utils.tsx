import type Queries from "@testing-library/dom/types/queries";
import type { RenderResult } from "@testing-library/react";
import { render } from "@testing-library/react";
import type { RequestHandler } from "msw";
import { setupServer } from "msw/node";
import { SWRConfig } from "swr";

import { SearchProvider } from "@/contexts/search/SearchContext";
import { ViewportProvider } from "@/contexts/viewport/ViewportContext";

import { withMockLegacyRouter, withMockRouter } from "./mocks";

export const Providers: React.ComponentType<{ children?: React.ReactNode }> = ({ children }) => {
  return withMockRouter(
    withMockLegacyRouter(
      <SWRConfig value={{ dedupingInterval: 0, provider: () => new Map() }}>
        <SearchProvider>
          <ViewportProvider>{children}</ViewportProvider>
        </SearchProvider>
      </SWRConfig>
    )
  );
};

const customRender = (ui: React.ReactElement, options = {}): RenderResult<typeof Queries, HTMLElement> => {
  return render(ui, { wrapper: Providers, ...options });
};

// re-export everything
export * from "./mocks";
export * from "@testing-library/react";

// override render method
export { render as originalRender, customRender as render };

export const reTestCase = {
  anyImage: expect.stringMatching(/^(data:image\/gif)|\.(png|webp|jpeg|jpg|svg)$/),
  anyWord: expect.stringMatching(/\w+/),
};

type ErrorComponentProps = {
  message?: string;
};
export const ErrorComponent: React.FC<ErrorComponentProps> = ({ message }) => {
  throw new Error(message);
};

export const setupMockServer = (...handlers: RequestHandler[]) => {
  const server = setupServer(...handlers);
  beforeAll(() => server.listen());
  afterEach(() => server.resetHandlers());
  afterAll(() => server.close());
  return server;
};
