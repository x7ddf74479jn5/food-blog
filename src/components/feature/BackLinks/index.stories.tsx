import type { ComponentMeta, ComponentStoryObj } from "@storybook/react";

import BackLinks from ".";
import { withRouterContext } from ".storybook/mocks/context";

export default {
  component: BackLinks,
  decorators: [
    (storyFn) => {
      return withRouterContext(storyFn(), { context: { pathname: "/" } });
    },
  ],
  title: "feature/BackLinks",
} as ComponentMeta<typeof BackLinks>;

export const Default: ComponentStoryObj<typeof BackLinks> = {
  args: {
    links: [
      { href: "test1", label: "ホームへ" },
      { href: "test2", label: "カテゴリー一覧へ" },
    ],
  },
};
