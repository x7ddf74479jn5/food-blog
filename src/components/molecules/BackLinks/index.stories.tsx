import type { ComponentMeta, ComponentStory } from "@storybook/react";

import BackLinks from ".";

export default {
  title: "molecules/BackLinks",
  component: BackLinks,
} as ComponentMeta<typeof BackLinks>;

const Template: ComponentStory<typeof BackLinks> = (args) => <BackLinks {...args} />;

export const Default = Template.bind({});
Default.args = {
  links: [
    { href: "test1", label: "ホームへ" },
    { href: "test2", label: "カテゴリー一覧へ" },
  ],
};
