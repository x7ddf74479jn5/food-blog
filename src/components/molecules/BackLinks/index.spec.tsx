import { render, screen } from "jest/test-utils";
import React from "react";
import renderer from "react-test-renderer";

import { getBackLinks, UrlTable } from "@/utils/paths/url";

import BackLinks from "../BackLinks";

describe("components/molecules/BackLinks", () => {
  const mockBackLinks = getBackLinks([UrlTable.home]);

  it("snapshot", () => {
    const tree = renderer.create(<BackLinks links={mockBackLinks} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("OK: 表示が正しい", () => {
    render(<BackLinks links={mockBackLinks} />);
    const anchor = screen.getByRole("link");
    expect(anchor).toHaveAttribute("href", UrlTable.home);
    expect(anchor).toHaveTextContent(mockBackLinks[0].label);
  });
});
