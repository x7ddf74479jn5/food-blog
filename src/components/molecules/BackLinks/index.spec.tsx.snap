// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/molecules/BackLinks snapshot 1`] = `
<ul
  className="flex flex-col space-y-2 align-top"
>
  <li>
    <button
      className="flex flex-row gap-2 items-center"
      onClick={[Function]}
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 320 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
        />
      </svg>
      前のページへ
    </button>
  </li>
  <li>
    <a
      className="flex flex-row gap-2 items-center"
      href="/"
      onClick={[Function]}
      onMouseEnter={[Function]}
    >
      <svg
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        strokeWidth="0"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 320 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
        />
      </svg>
      レシピ一覧へ
    </a>
  </li>
</ul>
`;
